pmd-core/src/main/java/net/sourceforge/pmd/benchmark/TimeTracker.java

```diff
@@ -0,0 +1,296 @@
+    public static void finishThread() {
+        if (!trackTime) {
+            return;
+        }
+        
+        finishOperation(0);
+        
+        // if using a mono-thread, we may not be empty...
+        if (TIMER_ENTRIES.get().isEmpty()) {
+            TIMER_ENTRIES.remove();
+        }
+    }
+    
+    /**
+     * Starts tracking an operation.
+     * @param category The category under which to track the operation.
+     */
+    public static TimedOperation startOperation(final TimedOperationCategory category) {
+        return startOperation(category, null);
+    }
+    
+    /**
+     * Starts tracking an operation.
+     * @param category The category under which to track the operation.
+     * @param label A label to be added to the category. Allows to differentiate measures within a single category.
+     */
+    public static TimedOperation startOperation(final TimedOperationCategory category, final String label) {
+        if (!trackTime) {
+            return NOOP_TIMED_OPERATION;
+        }
+        
+        TIMER_ENTRIES.get().add(new TimerEntry(category, label));
+        return new TimedOperationImpl();
+    }
+    
+    /**
+     * Finishes tracking an operation.
+     * @param extraDataCounter An optional additional data counter to track along the measurements.
+     *                         Users are free to track any extra value they want (ie: number of analyzed nodes,
+     *                         iterations in a loop, etc.)
+     */
+    /* default */ static void finishOperation(final long extraDataCounter) {
+        if (!trackTime) {
+            return;
+        }
+
+        final Queue<TimerEntry> queue = TIMER_ENTRIES.get();
+        final TimerEntry timerEntry = queue.remove();
+
+        // Compute if absent
+        TimedResult result = ACCUMULATED_RESULTS.get(timerEntry.operation);
+        if (result == null) {
+            ACCUMULATED_RESULTS.putIfAbsent(timerEntry.operation, new TimedResult());
+            result = ACCUMULATED_RESULTS.get(timerEntry.operation);
+        }
+
+        // Update counters and let next element on the stack ignore the time we spent
+        final long delta = result.accumulate(timerEntry, extraDataCounter);
+        if (!queue.isEmpty()) { 
+            queue.peek().inNestedOperations += delta;
+        }
+    }
+    
+    /**
+     * An entry in the open timers queue. Defines an operation that has started and hasn't finished yet.
+     */
+    private static class TimerEntry {
+        /* package */ final TimedOperationKey operation;
+        /* package */ final long start;
+        /* package */ long inNestedOperations = 0;
+        
+        /* package */ TimerEntry(final TimedOperationCategory category, final String label) {
+            this.operation = new TimedOperationKey(category, label);
+            this.start = System.nanoTime();
+        }
+
+        @Override
+        public String toString() {
+            return "TimerEntry for " + operation;
+        }
```

Comment:
Is this used?  Or was for debugging?  If keeping, should it include the other fields?

Reply:
Just debugging. Practicing Effective Java's Item 10. I chose this field alone to keep it uncluttered. When checking the contents of the stack one cares about the operations in it, rather than the numbers (which vary widely from run to run and specially under the presence of breakpoints)